### ISSUE TITLE
ARIA improvements for table-based views

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar#6526]
This ticket applies to all of FullCalendar's views except for [List view](https://fullcalendar.io/docs/list-view), which [has a separate ticket](https://github.com/fullcalendar/fullcalendar/issues/6525).

### Why nested tables?

FullCalendar utilizes nested tables. This is not very good for screen readers (pointed our [here](https://github.com/fullcalendar/fullcalendar/issues/4240#issuecomment-789198937), [here](https://github.com/fullcalendar/fullcalendar/issues/2536#issuecomment-789179551), [here](https://github.com/fullcalendar/fullcalendar/issues/3364#issuecomment-684764262), and many other tickets). Why does FullCalendar do this?

FullCalendar doesn't necessarily need tables-within-tables, but it DOES need the header of the table to have a different table DOM parent than the body. This is because the body often has scrollbars and scrolls independently of the header:

- For month view, this happens when large numbers of events vertically stack
- For timegrid view, the time slots scroll separately from the header and all-day section
- For any vertical resource view, the resources might scroll horizontally
- For resource-timeline view, the event area scrolls independently of the resources

It is not possible to achieve this with one table element. Multiple table elements need to live within multiple scroll-parents and they need to be synchronized.

Often times scrollbars are NOT needed. Why not simplify the DOM to a single table in this case? Because it is easier to develop and test a single codepath the renders a more predictable DOM structure.

FullCalendar will utilize multiple disconnected tables for the foreseeable future. We will likely move away from `<table>`-based markup and use ARIA roles for everything. But for now, we need to keep the current markup and merely add to it in order to prevent breaking DOM changes.

### The solution

Annotate REAL rows/cells that we want screen readers to read with plenty of [ARIA table roles](https://www.w3.org/TR/wai-aria-practices-1.1/#table_roles_states_props):

- `role="table"` (only on the root, to signal we are leaning on ARIA attributes for the children)
- `role="row"`
- `role="columnheader"`
- `role="rowheader"`
- `role="cell"`
- `role="presentation"` (for elements that DON'T behave as rows/cells but that unfortunately are `thead`/`tbody`/`tr`/`th`/`td` elements)

### Why not role="grid" ?

In order for a table to qualify as a ["grid" component](https://www.w3.org/TR/wai-aria-practices/#grid), it needs to internally manage focus within itself via a ["roving tabindex"](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex) ([example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/grid/dataGrids.html)).

There are plans to make this happen ([view ticket](https://github.com/fullcalendar/fullcalendar/issues/6527)). Until then, we'll declare a simpler "table" component.

### Special cases for timegrid

As mentioned [in this comment](https://github.com/fullcalendar/fullcalendar/issues/2536#issuecomment-703825166), the empty header cell at the top-left corner of timegrid view either needs to be `aria-hidden="true"` or have some sort of `aria-label` like "Time".

### Special cases for resource-timeline

The event pane of resource-timeline scrolls independently of the resources and their data cells. This fragmentation of the row across different parts of the DOM makes things difficult. You can explicitly define the relationship between the main row and the disconnected cell by using the [aria-owns](https://www.w3.org/TR/wai-aria-1.1/#aria-owns) attribute to simulate a child.

Unfortunately this will most likely result in a poorly-flowing tab order. [This ticket](https://github.com/fullcalendar/fullcalendar/issues/6527) will address it.

**Comment by danderson421** [comment]
Not sure this solution covers the requirements. The table structure itself is malformed.  th elements should be in thead. I don't believe simply specifying a role would satisfy the WGAC requirements. The required table structure is documented in the comments here: https://github.com/fullcalendar/fullcalendar/issues/6388

**Comment by danderson421** [comment]
The changes made in 5.1 look pretty good. It looks like there is one issue. According to the specification, presentation tables cannot have headers.


`<table role="presentation" class="fc-scrollgrid-sync-table" style="width: 688px;"><thead><tr><th class="fc-col-header-cell fc-day fc-day-sun">Sun</th><th class="fc-col-header-cell fc-day fc-day-mon">Mon</th><th class="fc-col-header-cell fc-day fc-day-tue">Tue</th><th class="fc-col-header-cell fc-day fc-day-wed">Wed</th><th class="fc-col-header-cell fc-day fc-day-thu">Thu</th><th class="fc-col-header-cell fc-day fc-day-fri">Fri</th><th class="fc-col-header-cell fc-day fc-day-sat">Sat</th></tr></thead>'


https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_table_01
